lib/net.c

```diff
@@ -95,125 +95,187 @@ randomize_rss_key(int guarantee_random_entropy)
-				port_id, ether_type);
-			ret = -1;
-			goto out;
-		}
+	if (!iface->rss) {
+		/*
+		 * If RSS is not supported, then data plane packets
+		 * could be assigned to RX queues that are serviced
+		 * by non-data plane blocks (e.g., LLS).
+		 */
+		G_LOG(NOTICE, "net: cannot use EtherType filters when RSS is not supported\n");
+		return -1;
 	}
 
-	ret = 0;
+	memset(&attr, 0, sizeof(struct rte_flow_attr));
+	attr.ingress = 1;
 
-out:
-	return ret;
-}
+	memset(action, 0, sizeof(action));
+	action[0].type = RTE_FLOW_ACTION_TYPE_QUEUE;
+	action[0].conf = &queue;
+	action[1].type = RTE_FLOW_ACTION_TYPE_END;
 
-/*
- * ntuple filters can only be used if supported by the NIC
- * (to steer matching packets) and if RSS is supported
- * (to steer non-matching packets elsewhere).
- */
-static inline bool
-hw_filter_ntuple_available(const struct gatekeeper_if *iface)
-{
-	return iface->hw_filter_ntuple && iface->rss;
+	memset(pattern, 0, sizeof(pattern));
+	pattern[0].type = RTE_FLOW_ITEM_TYPE_ETH;
+	memset(&eth_spec, 0, sizeof(struct rte_flow_item_eth));
+	memset(&eth_mask, 0, sizeof(struct rte_flow_item_eth));
+	eth_spec.type = rte_cpu_to_be_16(ether_type);
+	eth_mask.type = 0xFFFF;
+	pattern[0].spec = &eth_spec;
+	pattern[0].mask = &eth_mask;
+	pattern[1].type = RTE_FLOW_ITEM_TYPE_END;
+
+	ret = rte_flow_validate(iface->id, &attr, pattern, action, &error);
+	if (ret < 0) {
+		/*
+		 * A negative errno value was returned
+		 * (and also put in rte_errno).
+		 */
+		G_LOG(NOTICE, "net: could not validate EtherType flow: %s (%s)\n",
+			strerror(-ret), error.message);
+		return -1;
+	}
+
+	flow = rte_flow_create(iface->id, &attr, pattern, action, &error);
+	if (flow == NULL) {
+		/* rte_errno is set to a positive errno value. */
+		G_LOG(NOTICE, "net: could not create EtherType flow: %s (%s)\n",
+			strerror(rte_errno), error.message);
+		return -1;
+	}
+
+	return 0;
 }
 
 static int
-ipv4_ntuple_filter_add(struct gatekeeper_if *iface, rte_be32_t dst_ip_be,
+ipv4_flow_add(struct gatekeeper_if *iface, rte_be32_t dst_ip_be,
 	rte_be16_t src_port_be, rte_be16_t src_port_mask_be,
 	rte_be16_t dst_port_be, rte_be16_t dst_port_mask_be,
 	uint8_t proto, uint16_t queue_id)
 {
-	unsigned int i;
-	struct rte_eth_ntuple_filter filter_v4 = {
-		.flags = RTE_5TUPLE_FLAGS,
-		.dst_ip = dst_ip_be,
-		.dst_ip_mask = UINT32_MAX,
-		.src_ip = 0,
-		.src_ip_mask = 0,
-		.dst_port = dst_port_be,
-		.dst_port_mask = dst_port_mask_be,
-		.src_port = src_port_be,
-		.src_port_mask = src_port_mask_be,
-		.proto = proto,
-		.proto_mask = UINT8_MAX,
-		.tcp_flags = 0,
-		.priority = 1,
-		.queue = queue_id,
-	};
+	struct rte_flow_attr attr;
+	struct rte_flow_item pattern[3];
+	struct rte_flow_action action[2];
+	struct rte_flow *flow;
+	struct rte_flow_action_queue queue = { .index = queue_id };
+	struct rte_flow_item_eth eth_spec;
+	struct rte_flow_item_eth eth_mask;
+	struct rte_flow_item_ipv4 ip_spec;
+	struct rte_flow_item_ipv4 ip_mask;
+	struct rte_flow_error error;
```

Comment:
With the exception of the variable `pattern`, all these structs are being filled with fixed parameters, thus, it's better to initialize these structs at their definitions; like is being done with variable `queue` and was done with the variable `struct rte_eth_ethertype_filter filter` being removed by this patch.

This comment also applies to `ethertype_flow_add()`.

Reply:
Most of `pattern` can also be initialized during its declaration.